backend/utils/pubsub.js

```diff
@@ -0,0 +1,254 @@
+export { createClient, createMessage, NOTIFICATION_TYPE, REQUEST_TYPE, RESPONSE_TYPE }
+
+export function createNotification (
+  type: string,
+  data: {| contractID: string, socketID: number |}
+): string {
+  return JSON.stringify({ type, ...data })
+}
+
+export function createResponse (type: string, data: JSONType): string {
+  return JSON.stringify({ type, data })
+}
+
+/**
+ * Creates a pubsub server instance.
+ *
+ * @param {(http.Server|https.Server)} server - A Node.js HTTP/S server to attach to.
+ * @param {Object?} options
+ * {object?} clientHandlers - Custom handlers for socket events.
+ * {object?} messageHandlers - Custom handlers for different message types.
+ * {object?} serverHandlers - Custom handlers for server events.
+ *
+ * {number} backlog - The maximum length of the queue of pending connections.
+ * {boolean} clientTracking - Specifies whether or not to track clients.
+ * {Function} handleProtocols - A function which can be used to handle the WebSocket subprotocols.
+ * {number} maxPayload - The maximum allowed message size in bytes.
+ * {string} path - Accept only connections matching this path.
+ * {(boolean|object)} perMessageDeflate - Enables/disables per-message deflate.
+ * @returns {Object}
+ */
+export function createServer (httpServer: Object, options?: Object = {}): Object {
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = options.clientHandlers || {}
+  server.customServerHandlers = options.serverHandlers || {}
+  server.messageHandlers = { ...defaultMessageHandlers, ...options.messageHandlers }
+  server.pingInterval = undefined
+  server.subscribersByContractID = Object.create(null)
+
+  // Create and attach server-side WebSocket event listeners.
+  ;['close', 'connection', 'error', 'headers', 'listening'].forEach((name) => {
+    server.on(name, (...args) => {
+      console.log('[pubsub] Server event:', name)
+      try {
+        const customHandler = server.customServerHandlers[name]
+        const defaultHandler = defaultServerHandlers[name]
+        // Always call the default handler first.
+        if (defaultHandler) {
+          defaultHandler.call(server, ...args)
+        }
+        if (customHandler) {
+          customHandler.call(server, ...args)
+        }
+      } catch (error) {
+        logger(error)
+      }
+    })
+  })
+  // Setup a ping interval if required.
+  if (server.options.pingInterval > 0) {
+    server.pingInterval = setInterval(() => {
+      server.clients.forEach((client) => {
+        if (!client.isAlive) {
+          return client.terminate()
+        }
+        client.isAlive = false
+        client.ping()
+      })
+    }, server.options.pingInterval)
+  }
+  return server
+}
+
+const defaultOptions = {
+  clientTracking: true,
+  maxPayload: 6 * 1024 * 1024,
+  pingInterval: undefined
+}
+
+const generateSocketID = (() => {
+  let counter = 0
+
+  return () => counter++
```

Comment:
Consider a long running program that has a few 100k users. This value can become very big, especially after months or years of operation. That's problematic for a few reasons:

1. It will look quite ugly in the console.
2. In my tests, setting a variable to `Number.MAX_VALUE` causes that number to kind of break. You can no longer increment it or decrement it. Although it's unlikely the counter will ever reach this high, it's still a theoretical ugliness.

OTOH, something like `crypto.randomBytes(6).toString('hex')` is unlikely to ever create a conflict, will always be of the same length in output, and can be called indefinitely without worry.

Reply:
Eh, it shouldn't be an issue, even with a DDoS attack, let's leave it as is.

Reply:
I understand that a plein integer counter is not the most robust way to assign an ID to something. But it's very nice in development and testing. We can use real uuid's later in production.

Reply:
Using a counter is fine. I don't think we'll need to use UUIDs. If we need to associate sockets between backend and frontend, that might be useful, and we can use the approach we discussed on Slack (that of creating a `socketDescription` function that logs some data that is shared between frontend and backend, e.g. a username or a key hash, and only if logging such things is enabled or we're in development mode).